pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -672,6 +673,42 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, ru
 	return
 }
 
+func (m *kubeGenericRuntimeManager) preStopSidecars(pod *v1.Pod, containers []*kubecontainer.Container) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(containers))
+
+	for _, container := range containers {
+		go func(container *kubecontainer.Container) {
+			defer wg.Done()
+			gracePeriod := int64(minimumGracePeriodInSeconds)
+			switch {
+			case pod.DeletionGracePeriodSeconds != nil:
+				gracePeriod = *pod.DeletionGracePeriodSeconds
+			case pod.Spec.TerminationGracePeriodSeconds != nil:
+				gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
+			}
+			var containerSpec *v1.Container
+			if pod != nil {
+				if containerSpec = kubecontainer.GetContainerSpec(pod, container.Name); containerSpec == nil {
+					klog.Errorf("failed to get containerSpec %q(id=%q) in pod %q when pre-stopping sidecar container", container.Name, container.ID.String(), format.Pod(pod))
+				}
+			} else {
+				// Restore necessary information if one of the specs is nil.
+				restoredPod, restoredContainer, err := m.restoreSpecsFromContainerLabels(container.ID)
+				if err != nil {
+					klog.Errorf("failed to restore containerSpec %q(id=%q) in pod %q when pre-stopping sidecar container", container.Name, container.ID.String(), format.Pod(pod))
+				}
+				pod, containerSpec = restoredPod, restoredContainer
+			}
+
+			if containerSpec.Lifecycle != nil && containerSpec.Lifecycle.PreStop != nil && gracePeriod > 0 {
```

Comment:
this changes current behavior - 

here if grace period is 0, the pre stop hook will not be executed at all, while current container preStop always gets executed if defined, but wait for gracePeriod amount of time.